### PR TITLE
Add core files.

### PR DESCRIPTION
**Pull request by mattcollier**
Dependent PR: https://github.com/digitalbazaar/bedrock-authn-did-jwt/pull/1

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+// TODO: make responding to this event optional
+bedrock.events.on(
+  'bedrock-authn-did-jwt.config.secretStore', (strategy, callback) => {
+    strategy.setStore(api, callback);
+  });
+
+bedrock.events.on('bedrock-mongodb.ready', callback => {
+  async.waterfall([
+    callback => {
+      database.openCollections(['jwtSecrets'], callback);
```
Comment (dlongley) [inline lib/index.js]: Let's call this: `jwtSecretStore`.

File: lib/index.js [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        fields: {id: 1},
+        options: {unique: true, background: false}
+      }], callback);
+    }], callback);
+});
+
+/**
+ * Initialize a namespaced keystore.
+ *
+ * @param options the options to use
```
Comment (dlongley) [inline lib/index.js]: I think we usually put a colon `:` at the end of `the options to use`

File: lib/index.js [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+ *   ttlInSecs the JWT ttl
+ * @param callback(err, state) called once the operation completes.
+ */
+api.provision = function(options, callback) {
+  options = _.assign({}, options, {provision: true});
+  _getState(options, err => callback(err));
+};
+
+/**
+ * Create a JWT
```
Comment (dlongley) [inline lib/index.js]: Add period `.`

File: lib/index.js (hunk 1 above)
Comment (dlongley) [inline lib/index.js]: I see you used `key store` in some of the other comments... `jwtKeyStore` would also be an ok name, your call.

File: lib/index.js [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+ *
+ * @param options the options to use
+ *   namespace the namespace for the keystore
+ *   payload the payload to sign and include in the JWT
```
Comment (dlongley) [inline lib/index.js]: We put periods `.` at the end of these lines in our other docs, so let's do so here to be consistent.

File: lib/index.js [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+  async.auto({
+    getState: function(callback) {
+      return _getState({namespace: options.namespace}, callback);
+    },
+    create: ['getState', function(callback, results) {
+      var state = results.getState;
+      var nowInSecs = Math.floor(Date.now() / 1000);
+      var notAfter = (nowInSecs + state.tokenTtlInSecs);
+      var payload =
+        _.assign({}, options.payload, {exp: notAfter, iat: nowInSecs});
```
Comment (dlongley) [inline lib/index.js]: Style nit, keep function on same line as equals if possible (and also use `const`), change to:

```js
const payload = _.assign(
  {}, options.payload, {exp: notAfter, iat: nowInSecs});
```

File: lib/index.js [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+ */
+api.sign = function(options, callback) {
+  async.auto({
+    getState: function(callback) {
+      return _getState({namespace: options.namespace}, callback);
+    },
+    create: ['getState', function(callback, results) {
+      var state = results.getState;
+      var nowInSecs = Math.floor(Date.now() / 1000);
+      var notAfter = (nowInSecs + state.tokenTtlInSecs);
```
Comment (dlongley) [inline lib/index.js]: All of these should be `const`

File: lib/index.js [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+      var payload =
+        _.assign({}, options.payload, {exp: notAfter, iat: nowInSecs});
+      var token = jwt.sign(payload, state.key.data, {
```
Comment (dlongley) [inline lib/index.js]: `const token`

File: lib/index.js [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+
+      callback(null, token);
+    }]
+  }, function(err, results) {
+    callback(err, results ? results.create : null);
+  });
+};
+
+/**
+ * Verify a JWT
```
Comment (dlongley) [inline lib/index.js]: Add period `.`

File: lib/index.js [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+ *
+ * @param options the options to use
+ *   namespace the namespace for the keystore
+ *   token the token to be verified
```
Comment (dlongley) [inline lib/index.js]: Add colon `:` and periods `.`

File: lib/index.js [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+ *
+ * @param callback(err, state) called once the operation completes.
+ */
+api.verify = function(options, callback) {
+  async.auto({
+    get: function(callback) {
+      _getState({namespace: options.namespace}, callback);
+    },
+    verify: ['get', function(callback, results) {
+      var state = results.get;
```
Comment (dlongley) [inline lib/index.js]: `const`

File: lib/index.js [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+
+      try {
+        // parse header
+        var header = jwt.decode(options.token, {complete: true}).header;
```
Comment (dlongley) [inline lib/index.js]: `const header`

File: lib/index.js [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+
+        // get key that matches key ID
+        var key;
```
Comment (dlongley) [inline lib/index.js]: `let key`

File: lib/index.js [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+        if(header.kid === state.key.id) {
+          key = state.key;
+        } else if(state.previousKey && header.kid === state.previousKey.id) {
+          key = state.previousKey;
+        } else {
+          throw new Error('Invalid key identifier in token.');
+        }
+        // decode and verify the token
+        var token = jwt.verify(
```
Comment (dlongley) [inline lib/index.js]: `const token`

File: lib/index.js [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+      callback(null, token);
+    }]
+  }, function(err, results) {
+    callback(err, results ? results.verify : false);
+  });
+};
+
+function _get(id, callback) {
+  database.collections.jwtSecrets.findOne(
+    {id: database.hash(id)}, {state: true}, function(err, record) {
```
Comment (dlongley) [inline lib/index.js]: `(err, record) => {`

File: lib/index.js [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+  var record = {
+    id: database.hash(id),
+    meta: {
+      created: now,
+      updated: now
+    },
+    state: state
+  };
+  database.collections.jwtSecrets.insert(
+    record, database.writeOptions, function(err, result) {
```
Comment (dlongley) [inline lib/index.js]: `(err, result) => {`

File: lib/index.js [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+          'NotFound',
+          {proof: id, httpStatusCode: 404, 'public': true}));
+      }
+      callback(null, record.state, record.meta);
+    });
+}
+
+function _insert(id, state, callback) {
+  var now = Date.now();
+  var record = {
```
Comment (dlongley) [inline lib/index.js]: `const`

File: lib/index.js [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+    });
+}
+
+function _update(id, state, options, callback) {
+  if(typeof options === 'function') {
+    callback = options;
+    options = {};
+  }
+  options = options || {};
+  var query = _.assign({}, options.query || {});
```
Comment (dlongley) [inline lib/index.js]: `const`

File: lib/index.js [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+  query.id = database.hash(id);
+  database.collections.jwtSecrets.update(
+    query, {
+      $set: {
+        'meta.updated': Date.now(),
+        'state': state
+      }
+    }, database.writeOptions, function(err, result) {
+      if(err) {
+        logger.error('Mongo error when trying to update proof state.', err);
```
Comment (dlongley) [inline lib/index.js]: `proof state` => `key state`

File: lib/index.js [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+        return callback(new BedrockError(
+          'Failed to update proof state due to internal error.',
```
Comment (dlongley) [inline lib/index.js]: `proof state` => `key state`

File: lib/index.js [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+      state.key = {
+        id: '' + nowInSecs,
+        algorithm: state.previousKey.algorithm,
+        data: _generateHmacKey(),
+        created: nowInSecs,
+        expires: nowInSecs + state.tokeTtlInSecs +
+          state.tokenClockToleranceInSecs
+      };
+      logger.verbose('[jwt-mongodb] recycling HMAC key...');
+      _update(options.algorithm, state, {
```
Comment (dlongley) [inline lib/index.js]: `options.algorithm` => `options.namespace`?

File: lib/index.js [hunk 21]
```diff
@@ -0,0 +1,333 @@
...
+        'meta.updated': Date.now(),
+        'state': state
+      }
+    }, database.writeOptions, function(err, result) {
+      if(err) {
+        logger.error('Mongo error when trying to update proof state.', err);
+        return callback(new BedrockError(
+          'Failed to update proof state due to internal error.',
+          'InternalError',
+          {proof: id, httpStatusCode: 500, 'public': true},
```
Comment (dlongley) [inline lib/index.js]: `proof: id` => `namespace: id`

File: lib/index.js [hunk 22]
```diff
@@ -0,0 +1,333 @@
...
+/**
+ * Gets the cached proof state information, retrieving it from the database
+ * and recycling HMAC keys as needed.
+ *
+ * @param options the options to use
+ *   algorithm the JWT signing algorithm to use
+ *   clockToleranceInSecs the clockTolerance to use in during JWT verification
+ *   namespace the namespace to use
+ *   provision true for provisioning operation
+ *   ttlInSecs the JWT ttl
```
Comment (dlongley) [inline lib/index.js]: Add punctuation

File: lib/index.js [hunk 23]
```diff
@@ -0,0 +1,333 @@
...
+ * @param callback(err, state) called once the operation completes.
+ */
+function _getState(options, callback) {
+  var done = callback;
+  async.auto({
+    get: function(callback) {
+      // use cached proof state if available, expiration checked later
```
Comment (dlongley) [inline lib/index.js]: Blanket replace `proof state` with `state` or `key state` in comments as you see fit.

File: lib/index.js [hunk 24]
```diff
@@ -0,0 +1,333 @@
...
+        key: {
+          id: '' + nowInSecs,
+          algorithm: options.algorithm,
+          data: _generateHmacKey(),
+          created: nowInSecs,
+          expires: nowInSecs + options.ttlInSecs + options.clockToleranceInSecs
+        }
+      };
+      _insert(options.namespace, state, function(err) {
+        if(err && err.name === 'DuplicateState') {
```
Comment (dlongley) [inline lib/index.js]: Let's call this `DuplicateRecord`.

File: lib/index.js [hunk 25]
```diff
@@ -0,0 +1,333 @@
...
+    },
+    state: state
+  };
+  database.collections.jwtSecrets.insert(
+    record, database.writeOptions, function(err, result) {
+      if(err) {
+        if(database.isDuplicateError(err)) {
+          return callback(new BedrockError(
+            'The proof state is a duplicate and could not be added.',
+            'DuplicateProofState', {
```
Comment (dlongley) [inline lib/index.js]: `DuplicateProofState` => `DuplicateRecord`

File: lib/index.js [hunk 26]
```diff
@@ -0,0 +1,333 @@
...
+      updated: now
+    },
+    state: state
+  };
+  database.collections.jwtSecrets.insert(
+    record, database.writeOptions, function(err, result) {
+      if(err) {
+        if(database.isDuplicateError(err)) {
+          return callback(new BedrockError(
+            'The proof state is a duplicate and could not be added.',
```
Comment (dlongley) [inline lib/index.js]: `proof state` => `key state`

**Comment by mattcollier** [comment]
Implemented in a different branch.

**Pull request closed without merging**